homeassistant/config_entries.py

```diff
@@ -1794,7 +1792,15 @@ async def async_reload(self, entry_id: str) -> bool:
         # attempts.
         entry.async_cancel_retry_setup()
 
-        async with entry.reload_lock:
+        if entry.domain not in self.hass.config.components:
+            # If the component is not loaded, just load it as
+            # the config entry will be loaded as well. We need
+            # to do this before holding the lock to avoid a
+            # deadlock.
+            await async_setup_component(self.hass, entry.domain, self._hass_config)
+            return entry.state is ConfigEntryState.LOADED
```

Comment:
This is exactly what `self.async_setup(entry_id)` does, but we need to do it without holding the lock since `async_setup_component` will hold the lock